doc/md/fundamentals/2-types/14-subtyping.md

```diff
@@ -4,121 +4,347 @@ sidebar_position: 14
+[Immutable arrays (`[T]`)](https://internetcomputer.org/docs/motoko/fundamentals/types/immutable-arrays) support covariant subtyping, meaning if `T <: U`, then `[T] <: [U]`.
+
+If `T <: U`, then an (immutable) array of type `[T]` can be used as an array of type `[U]`.
 
 ```motoko no-repl
-let nums : [Nat] = [1, 2, 3];
-let numsAsInts : [Int] = nums;  // Allowed, since `Nat <: Int`
+let nats : [Nat] = [1, 2, 3];
+let ints : [Int] = nats;  // Allowed, since `Nat <: Int` we also have `[Nat] <: [Int]`.
 ```
 
 ## Mutable arrays
 
-[Mutable arrays (`var [T]`)](https://internetcomputer.org/docs/motoko/fundamentals/types/mutable-arrays) do not support subtyping due to type safety risks. Allowing `var [T1] <: var [T2]` could lead to runtime type mismatches.
+[Mutable arrays](https://internetcomputer.org/docs/motoko/fundamentals/types/mutable-arrays) of the form `[var T]` do not support interesting subtyping.
+The mutable array constructor `[var T]` is invariant in `T`.
+
+This means that `[var T]` is a subtype of `[var U]` only if `T` and `U` are equivalent types, that is, both `T <: U` and `U <: T` must hold.
+
+Allowing `[var T] <: [var U]` whenever `T <: U` (without requiring `U <: T`) would be unsafe, as it could allow writing values of type `U` into an array that only expects values of type `T`.
 
 ```motoko no-repl
-let nums :  [var Nat] = [var 1, 2, 3];
-let numsAsInts :  [var Int] = nums;  // Not allowed because mutable arrays are invariant
+let nats : [var Nat] = [var 1, 2, 3];
+let ints : [var Int] = nats;  // Not allowed, because `[var Nat] </: [var Int]`.
 ```
 
 ## Functions
 
-[Function](https://internetcomputer.org/docs/motoko/fundamentals/types/functions) subtyping follows two rules:
+[Functions](https://internetcomputer.org/docs/motoko/fundamentals/types/functions) also support subtyping.
+A function type `T1 -> T2` is a subtype of another function type `U1 -> U2` provided that:
 
-1. Parameter types must be contravariant: `T2 <: T1` means `T1 -> R <: T2 -> R`
-2. Return types must be covariant: `R1 <: R2` means `T -> R1 <: T -> R2`
+1. The argument types are related in the opposite direction (`U1 <: T1` ).
+2. The return types are related in the same direction (`T2 <: U2`).
 
-```motoko no-repl
-func parent(x : Int) : Nat { 10 };
-func child(x : Nat) : Int { 20 };
+This means that function subtyping is contravariant in the arguments of the function and covariant in the results.
+
+This might seems confusing at first, but the intuition is quite simple.
+A function can be used at a supertype provided that type requires a more specific argument and produces a more general result.
+This ensures that the orginal function, when used at the supertype,
+will only receive arguments it can consume (since `U1 <: T1`) and produce results than are expected (since `T2 <: U2`).
```

Comment:
```suggestion
Function subtyping is contravariant in the argument types and covariant in the return type.

A function of type `T1 → T2` can be used where a function of type `U1 → U2` is expected only if `U1 <: T1` (the expected argument is more specific) and `T2 <: U2` (the actual result is more generic).

This ensures safety because the function will never be passed arguments it can’t handle, and it will always produce results that meet the expectations of the surrounding context.
```